### ISSUE TITLE
Key reuse for encryption and authentication

### ISSUE THREAD

**Issue by nlordell** [WalletConnect/walletconnect-monorepo#239]
It looks like the key that is being used for encrypting the message payloads being sent over the WebSocket is the same as the key being used for the HMAC authentication of the encrypted payload.

From what I gathered the AEAD scheme being used is Authenticated Encryption with AES-CBC and HMAC-SHA2 (`AEAD_AES_256_CBC_HMAC_SHA_256`) from the standard here: https://tools.ietf.org/html/draft-mcgrew-aead-aes-cbc-hmac-sha2-05. In the standard it is explicitly stated that the [encryption key and the mac key must not overlap](https://tools.ietf.org/html/draft-mcgrew-aead-aes-cbc-hmac-sha2-05#section-2.1):
> When generating the secondary keys from K, MAC_KEY and ENC_KEY MUST NOT overlap.

They are however the same keys in the current implementation. I was wondering if you could provide clarification as to why the implementation is different from the standard.

**Comment by pedrouid** [comment]
Hi @nlordell, thanks for writing about it. This was a mistake done early on the protocol development that was later pointed out but we've decided to keep it as is to not break backwards compatibility as it was still gaining adoption. 

The next major version of the protocol will introduce many breaking changes which will also include a better cryptographic scheme for the authenticated encryption.

I already started working on this in parallel in this repo which includes ECIES that can be used to derive a shared secret between peer's public keys.

https://github.com/pedrouid/eccrypto-js/blob/master/src/ecies.ts
